NetKet/Utils/array_search.hpp

```diff
@@ -0,0 +1,31 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_ARRAYSEARCH_HPP
+#define NETKET_ARRAYSEARCH_HPP
+
+namespace netket {
+// Finds the smallest non zero element of an array
+template <class ForwardIterator>
+ForwardIterator min_nonzero_elem(ForwardIterator first, ForwardIterator last) {
+  if (first == last) return last;
+  ForwardIterator smallest = first;
+
+  while (++first != last)
+    if (*first < *smallest and *first != 0) smallest = first;
```

Comment:
Using `and` and `or` in place of `&&` and `||` is considered bad practice, doesn't work on MSVC, and was removed in `C++17`. Could you, please, replace `and` with `&&`?